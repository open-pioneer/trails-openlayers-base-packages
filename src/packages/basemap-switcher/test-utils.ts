// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { MapConfig, MapConfigProvider, MapRegistry } from "@open-pioneer/map";
import TileLayer from "ol/layer/Tile";
import OSM from "ol/source/OSM";
import Stamen from "ol/source/Stamen";
import { createService } from "@open-pioneer/test-utils/services";
import { MapRegistryImpl } from "@open-pioneer/map/MapRegistryImpl";
import { PackageContextProviderProps } from "@open-pioneer/test-utils/react";
import { screen, waitFor } from "@testing-library/react";

// used to avoid a "ResizeObserver is not defined" error
import ResizeObserver from "resize-observer-polyfill";
global.ResizeObserver = ResizeObserver;

export interface SimpleMapOptions {
    center?: { x: number; y: number };
    zoom?: number;
}

export const MAP_ID = "test";

export async function waitForMapMount(parentTestId = "base") {
    return await waitFor(async () => {
        const domElement = await screen.findByTestId(parentTestId);
        const container = domElement.querySelector(".ol-viewport");
        if (!container) {
            throw new Error("map not mounted");
        }
        return domElement;
    });
}
export async function setupMap(options?: SimpleMapOptions) {
    const mapId = "test";
    const mapConfig: MapConfig = {
        initialView: {
            kind: "position",
            center: options?.center ?? { x: 847541, y: 6793584 },
            zoom: options?.zoom ?? 10
        },
        projection: "EPSG:3857",
        layers: [
            {
                id: "b-1",
                title: "OSM",
                isBaseLayer: true,
                visible: false,
                layer: new TileLayer({
                    source: new OSM()
                })
            },
            {
                id: "b-2",
                title: "Toner",
                isBaseLayer: true,
                visible: true,
                layer: new TileLayer({
                    source: new Stamen({ layer: "toner" })
                })
            }
        ]
    };

    const registry = await createService(MapRegistryImpl, {
        references: {
            providers: [new MapConfigProviderImpl(mapId, mapConfig)]
        }
    });

    return { mapId, registry };
}

export function createPackageContextProviderProps(
    service: MapRegistry
): PackageContextProviderProps {
    return {
        services: {
            "map.MapRegistry": service
        }
    };
}

class MapConfigProviderImpl implements MapConfigProvider {
    mapId = "default";
    mapConfig: MapConfig;

    constructor(mapId: string, mapConfig?: MapConfig | undefined) {
        this.mapId = mapId;
        this.mapConfig = mapConfig ?? {};
    }

    getMapConfig(): Promise<MapConfig> {
        return Promise.resolve(this.mapConfig);
    }
}
