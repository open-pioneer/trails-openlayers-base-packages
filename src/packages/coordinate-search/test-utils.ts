// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { act, fireEvent } from "@testing-library/react";

export function showDropdown(projSelect: HTMLElement) {
    // open dropdown to include options in snapshot; react-select creates list of options in dom after opening selection
    act(() => {
        fireEvent.keyDown(projSelect, { key: "ArrowDown" });
    });
}

export function getCurrentOptions() {
    return Array.from(
        document.getElementsByClassName("coordinate-input-select__option")
    ) as HTMLElement[];
}

export function getCurrentOptionValues(options: HTMLElement[]) {
    const values: (string | null)[] = [];
    for (const opt of options) {
        values.push(opt.textContent);
    }
    return values;
}

export function getClearButton(coordinateSearchGroup: Element) {
    const buttonDiv = coordinateSearchGroup.querySelector(".chakra-input__right-element");
    if (!buttonDiv) {
        throw new Error("buttons not rendered");
    }
    const clearButton = buttonDiv.querySelector(".coordinate-input-clear-button");
    if (!clearButton) {
        throw new Error("clear button not rendered");
    }
    return clearButton;
}

export function getCopyButton(coordinateSearchGroup: Element) {
    const buttonDiv = coordinateSearchGroup.querySelector(".chakra-input__right-element");
    if (!buttonDiv) {
        throw new Error("buttons not rendered");
    }
    const copyButton = buttonDiv.querySelector(".coordinate-input-copy-button");
    if (!copyButton) {
        throw new Error("copy button not rendered");
    }
    return copyButton;
}
