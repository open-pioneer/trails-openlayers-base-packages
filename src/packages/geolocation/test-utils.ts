// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import OlMap from "ol/Map";
import { Circle, Fill, Stroke, Style } from "ol/style";
import { vi } from "vitest";
import { GeolocationController } from "./GeolocationController";

export function mockSuccessGeolocation(coords: number[]) {
    vi.spyOn(navigator, "geolocation", "get").mockReturnValue({
        clearWatch() {},
        getCurrentPosition() {},
        watchPosition(success) {
            setTimeout(() => {
                success({
                    coords: {
                        latitude: coords[0],
                        longitude: coords[1],
                        accuracy: 2500,
                        altitude: null,
                        altitudeAccuracy: null,
                        heading: null,
                        speed: null
                    }
                } as GeolocationPosition);
            }, 1);

            return 123;
        }
    } satisfies Partial<Geolocation>);
}

export function mockErrorGeolocation() {
    vi.spyOn(navigator, "geolocation", "get").mockReturnValue({
        clearWatch() {},
        getCurrentPosition() {},
        watchPosition(success, error) {
            setTimeout(() => {
                error?.({
                    code: 2,
                    message: "POSITION_UNAVAILABLE",
                    PERMISSION_DENIED: 1,
                    POSITION_UNAVAILABLE: 2,
                    TIMEOUT: 3
                });
            }, 1);

            return 123;
        }
    } satisfies Partial<Geolocation>);
}

export function setup() {
    const olMap = new OlMap();
    return {
        olMap,
        controller: new GeolocationController(olMap, (error) => {
            console.error("Unexpected error", error);
        })
    };
}

export function setupWithCustomProperties() {
    const olMap = new OlMap();
    const maxZoom: number = getCustomMaxZoom();
    const positionFeatureStyle: Style = getCustomPositionStyle();
    const accuracyFeatureStyle: Style = getCustomAccuracyStyle();
    const trackingOptions: PositionOptions = getCustomTrackingOptions();

    const controller = new GeolocationController(
        olMap,
        (error) => {
            console.error("Unexpected error", error);
        },
        trackingOptions
    );
    controller.setMaxZoom(maxZoom);
    controller.setPositionFeatureStyle(positionFeatureStyle);
    controller.setAccuracyFeatureStyle(accuracyFeatureStyle);
    return {
        olMap,
        controller
    };
}

export function getCustomMaxZoom() {
    return 20;
}
export function getCustomPositionStyle() {
    return new Style({
        image: new Circle({
            fill: new Fill({
                color: "rgba(255,255,255,0.4)"
            }),
            stroke: new Stroke({
                color: "#3399CC",
                width: 1.25
            }),
            radius: 5
        }),
        fill: new Fill({
            color: "rgba(255,255,255,0.4)"
        }),
        stroke: new Stroke({
            color: "#3399CC",
            width: 1.25
        })
    });
}

export function getCustomAccuracyStyle() {
    return new Style({
        image: new Circle({
            fill: new Fill({
                color: "rgba(255,255,255,0.4)"
            }),
            stroke: new Stroke({
                color: "#3399CC",
                width: 1.25
            }),
            radius: 5
        }),
        fill: new Fill({
            color: "rgba(255,255,255,0.4)"
        }),
        stroke: new Stroke({
            color: "#3399CC",
            width: 1.25
        })
    });
}

function getCustomTrackingOptions() {
    return {
        "enableHighAccuracy": true,
        "timeout": 20,
        "maximumAge": 50
    };
}
