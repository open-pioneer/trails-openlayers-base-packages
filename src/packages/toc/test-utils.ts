// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { screen } from "@testing-library/react";

export async function findToc() {
    const tocDiv = await screen.findByTestId("toc");
    return tocDiv;
}
