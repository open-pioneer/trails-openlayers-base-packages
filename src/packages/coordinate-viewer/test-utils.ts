// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { OlMapConfigurationProvider } from "@open-pioneer/experimental-ol-map";
import { OlMapRegistry } from "@open-pioneer/experimental-ol-map/services";
import { ServiceOptions } from "@open-pioneer/runtime";
import { PackageContextProviderProps } from "@open-pioneer/test-utils/react";
import { createService } from "@open-pioneer/test-utils/services";
import { waitFor, screen } from "@testing-library/react";
import { MapOptions } from "ol/Map";
import View from "ol/View";
import TileLayer from "ol/layer/Tile";
import OSM from "ol/source/OSM";

// used to avoid a "ResizeObserver is not defined" error
import ResizeObserver from "resize-observer-polyfill";
global.ResizeObserver = ResizeObserver;

export interface SimpleMapOptions {
    center?: [number, number];
    zoom?: number;
}

export const MAP_ID = "test";

export async function waitForMapMount(parentTestId = "base") {
    return await waitFor(async () => {
        const domElement = await screen.findByTestId(parentTestId);
        const container = domElement.querySelector(".ol-viewport");
        if (!container) {
            throw new Error("map not mounted");
        }
        return domElement;
    });
}

export async function setupMap(options?: SimpleMapOptions) {
    const mapOptions: MapOptions = {
        view: new View({
            projection: "EPSG:3857",
            center: options?.center ?? [847541, 6793584],
            zoom: options?.zoom ?? 10
        }),
        layers: [
            new TileLayer({
                source: new OSM(),
                properties: { title: "OSM" }
            })
        ]
    };

    const mapConfigProvider = await createService(MapConfigProvider, {
        properties: {
            mapOptions: mapOptions,
            mapId: MAP_ID
        }
    });
    const registry = await createService(OlMapRegistry, {
        references: {
            providers: [mapConfigProvider]
        }
    });

    return { mapId: MAP_ID, registry };
}

export function createPackageContextProviderProps(
    service: OlMapRegistry
): PackageContextProviderProps {
    return {
        services: {
            "ol-map.MapRegistry": service
        }
    };
}

class MapConfigProvider implements OlMapConfigurationProvider {
    mapId = "default";
    mapOptions: MapOptions = {};

    constructor(options: ServiceOptions) {
        if (options.properties.mapOptions) {
            this.mapOptions = options.properties.mapOptions as MapOptions;
        }
        if (options.properties.mapId) {
            this.mapId = options.properties.mapId as string;
        }
    }

    getMapOptions(): Promise<MapOptions> {
        return Promise.resolve(this.mapOptions);
    }
}
