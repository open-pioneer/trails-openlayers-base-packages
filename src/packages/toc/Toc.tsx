// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { Box, Text, FormControl, FormLabel } from "@open-pioneer/chakra-integration";
import { FC } from "react";
import { CommonComponentProps, useCommonComponentProps } from "@open-pioneer/react-utils";
import { BasemapSwitcher, BasemapSwitcherProps } from "@open-pioneer/basemap-switcher";
import { useIntl } from "open-pioneer:react-hooks";

export interface TocProps extends CommonComponentProps {
    /**
     * The id of the map.
     */
    mapId: string;

    /**
     * Defines whether the basemap switcher is shown in the toc.
     * Defaults to false.
     */
    hideBasemapSwitcher?: boolean;

    /**
     * Properties for the embedded basemap switcher.
     * Property "mapId" is not applied.
     */
    basemapSwitcherProps?: Omit<BasemapSwitcherProps, "mapId">;
}

export const Toc: FC<TocProps> = (props) => {
    const intl = useIntl();

    const { mapId, hideBasemapSwitcher = false, basemapSwitcherProps } = props;
    const { containerProps } = useCommonComponentProps("toc", props);
    const basemapsLabel = intl.formatMessage({ id: "basemapsLabel" });
    const tocTitel = intl.formatMessage({ id: "tocTitel" });

    return (
        <Box {...containerProps}>
            {/*TODO: remove header as it should be added by an app developer if needed? */}
            <Box
                className="toc-header"
                padding={2}
                backgroundColor="var(--chakra-colors-blackAlpha-500)"
            >
                <Text as="b">{tocTitel}</Text>
            </Box>
            {hideBasemapSwitcher || (
                <Box className="toc-content" padding={2}>
                    <FormControl>
                        <FormLabel ps={1}>
                            <Text as="b">{basemapsLabel}:</Text>
                        </FormLabel>
                        <BasemapSwitcher {...basemapSwitcherProps} mapId={mapId}></BasemapSwitcher>
                    </FormControl>
                </Box>
            )}
        </Box>
    );
};
