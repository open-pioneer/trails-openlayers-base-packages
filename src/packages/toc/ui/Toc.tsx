// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { reactive, reactiveMap } from "@conterra/reactivity-core";
import { BasemapSwitcher, BasemapSwitcherProps } from "@open-pioneer/basemap-switcher";
import { Box, Flex, Spacer, Text } from "@open-pioneer/chakra-integration";
import { MapModel, MapModelProps, useMapModel } from "@open-pioneer/map";
import {
    CommonComponentProps,
    SectionHeading,
    TitledSection,
    useCommonComponentProps
} from "@open-pioneer/react-utils";
import { useIntl } from "open-pioneer:react-hooks";
import { FC, useEffect, useId, useMemo, useRef } from "react";
import { TocItem, TocModel, TocModelProvider, TocWidgetOptions } from "../model/TocModel";
import { TopLevelLayerList } from "./LayerList/LayerList";
import { Tools } from "./Tools";

/**
 * Props supported by the {@link Toc} component.
 */
export interface TocProps extends CommonComponentProps, MapModelProps {
    /**
     * Defines whether the tool component is shown in the toc.
     *
     * Defaults to `false`.
     */
    showTools?: boolean;

    /**
     * Properties for the embedded tool component.
     */
    toolsConfig?: ToolsConfig;

    /**
     * Defines whether the basemap switcher is shown in the toc.
     *
     * Defaults to `true`.
     */
    showBasemapSwitcher?: boolean;

    /**
     * Properties for the embedded basemap switcher.
     *
     * Property `mapId` is not applied (the basemap switcher uses the same map as the).
     */
    basemapSwitcherProps?: Omit<BasemapSwitcherProps, "mapId">;

    /**
     * If `true`, groups in the toc can be collapsed and expanded.
     * This property should only be `true` if the map actually contains layer groups.
     *
     * Defaults to `false`.
     */
    collapsibleGroups?: boolean;

    /**
     * If `true` groups in the toc are collapsed initially.
     *
     * Defaults to `false`. If {@link collapsibleGroups} is `false` this property should also be `false`. Otherwise only the top level layers will appear in the toc.
     */
    initiallyCollapsed?: boolean;

    /**
     * Show the parent layers when a child layer is made visible.
     *
     * Defaults to `true`.
     */
    autoShowParents?: boolean;
}

/**
 * Props supported by the {@link Tools} component.
 */
export interface ToolsConfig {
    /**
     * Optional property to show the `hide all layers` entry.
     *
     * Defaults to `true`.
     */
    showHideAllLayers?: boolean;

    /**
     * Optional property to show the `collapse all groups` entry.
     *
     * Defaults to `true`. Only applicable if {@link TocProps.collapsibleGroups} is `true`.
     */
    showCollapseAllGroups?: boolean;
}

const PADDING = 2;

/**
 * Displays the layers of the configured map.
 */
export const Toc: FC<TocProps> = (props: TocProps) => {
    const intl = useIntl();
    const { containerProps } = useCommonComponentProps("toc", props);
    const state = useMapModel(props);

    let content: JSX.Element | null;
    switch (state.kind) {
        case "loading":
            content = null;
            break;
        case "rejected":
            content = <Text className="toc-error">{intl.formatMessage({ id: "error" })}</Text>;
            break;
        case "resolved": {
            const map = state.map;
            content = <TocContent {...props} map={map} />;
            break;
        }
    }

    return (
        <Flex {...containerProps} direction="column" gap={PADDING}>
            {content}
        </Flex>
    );
};

/** This component is rendered once we have a reference to the loaded map model. */
function TocContent(props: TocProps & { map: MapModel }) {
    const {
        map,
        showTools = false,
        toolsConfig,
        showBasemapSwitcher = true,
        basemapSwitcherProps
    } = props;
    const intl = useIntl();
    const model = useTocModel(props);
    const basemapsHeadingId = useId();
    const basemapSwitcher = showBasemapSwitcher && (
        <Box className="toc-basemap-switcher">
            <TitledSection
                title={
                    <SectionHeading id={basemapsHeadingId} size={"sm"} mb={PADDING}>
                        {intl.formatMessage({ id: "basemapsLabel" })}
                    </SectionHeading>
                }
            >
                <BasemapSwitcher
                    map={map}
                    aria-labelledby={basemapsHeadingId}
                    {...basemapSwitcherProps}
                />
            </TitledSection>
        </Box>
    );

    const layerList = (
        <Box className="toc-operational-layers">
            <TitledSection
                title={
                    <SectionHeading size={"sm"} mb={2}>
                        <Flex>
                            <Text my={3}>
                                {intl.formatMessage({
                                    id: "operationalLayerLabel"
                                })}
                            </Text>
                            <Spacer />
                            {showTools && <Tools map={map} {...toolsConfig} />}
                        </Flex>
                    </SectionHeading>
                }
            >
                <TopLevelLayerList
                    map={map}
                    aria-label={intl.formatMessage({ id: "operationalLayerLabel" })}
                />
            </TitledSection>
        </Box>
    );

    return (
        <TocModelProvider value={model}>
            {basemapSwitcher}
            {layerList}
        </TocModelProvider>
    );
}

function useTocModel(props: TocProps): TocModel {
    const initialProps = useRef(props);
    const { model, options } = useMemo(() => {
        const options = reactive<TocWidgetOptions>(
            createOptions(
                initialProps.current.autoShowParents,
                initialProps.current.collapsibleGroups,
                initialProps.current.initiallyCollapsed
            )
        );

        // Indexed by layerId
        const items = reactiveMap<string, TocItem>();
        const model: TocModel = {
            get options() {
                return options.value;
            },
            getItem(layerId: string): TocItem | undefined {
                return items.get(layerId);
            },
            getItems(): TocItem[] {
                return Array.from(items.values());
            },
            registerItem(item: TocItem): void {
                if (items.has(item.layerId)) {
                    throw new Error(`Item with layerId '${item.layerId}' already registered.`);
                }
                items.set(item.layerId, item);
            },
            unregisterItem(item: TocItem): void {
                if (items.get(item.layerId) !== item) {
                    throw new Error(`Item with layerId '${item.layerId}' not registered.`);
                }
                items.delete(item.layerId);
            }
        };
        return { model, options };
    }, []);

    // Sync props to model
    useEffect(() => {
        options.value = createOptions(
            props.autoShowParents,
            props.collapsibleGroups,
            props.initiallyCollapsed
        );
    }, [options, props.autoShowParents, props.collapsibleGroups, props.initiallyCollapsed]);

    return model;
}

function createOptions(
    autoShowParents?: boolean | undefined,
    collapsibleGroups?: boolean | undefined,
    isCollapsed?: boolean | undefined
): TocWidgetOptions {
    return {
        autoShowParents: autoShowParents ?? true,
        collapsibleGroups: collapsibleGroups ?? false,
        initiallyCollapsed: isCollapsed ?? false
    };
}
