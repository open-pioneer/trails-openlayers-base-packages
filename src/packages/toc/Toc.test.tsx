// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { createServiceOptions, setupMap } from "@open-pioneer/map-test-utils";
import { PackageContextProvider } from "@open-pioneer/test-utils/react";
import { render, screen, waitFor } from "@testing-library/react";
import TileLayer from "ol/layer/Tile";
import { expect, it } from "vitest";
import { Toc } from "./Toc";

const BASEMAP_SWITCHER_CLASS = ".basemap-switcher";

it("should successfully create a toc component", async () => {
    const { mapId, registry } = await setupMap({
        layers: [
            {
                title: "Base layer",
                id: "base-layer",
                olLayer: new TileLayer({}),
                isBaseLayer: true
            },
            {
                title: "Layer 1",
                id: "layer-1",
                olLayer: new TileLayer({})
            },
            {
                title: "Layer 2",
                id: "layer-2",
                olLayer: new TileLayer({})
            }
        ]
    });
    await registry.expectMapModel(mapId);
    const injectedServices = createServiceOptions({ registry });

    render(
        <PackageContextProvider services={injectedServices}>
            <Toc mapId={mapId} data-testid="toc" />
        </PackageContextProvider>
    );

    const tocDiv = await findToc();
    const { basemapSelect } = await waitForBasemapSwitcher(tocDiv!);
    await waitFor(() => {
        const option = basemapSelect.options[0];
        if (!option || option.innerText !== "Base layer") {
            throw new Error("expected basemap switcher to contain the Base layer option");
        }
    });
    expect(tocDiv).toMatchSnapshot();
});

it("should successfully create a toc component with additional css classes", async () => {
    const { mapId, registry } = await setupMap();
    await registry.expectMapModel(mapId);
    const injectedServices = createServiceOptions({ registry });

    render(
        <PackageContextProvider services={injectedServices}>
            <Toc mapId={mapId} className="test" data-testid="toc" />
        </PackageContextProvider>
    );

    const tocDiv = await findToc();
    expect(tocDiv.classList.contains("test")).toBe(true);
});

it("should embed the basemap switcher by default", async () => {
    const { mapId, registry } = await setupMap();
    await registry.expectMapModel(mapId);
    const injectedServices = createServiceOptions({ registry });

    render(
        <PackageContextProvider services={injectedServices}>
            <Toc mapId={mapId} data-testid="toc"></Toc>
        </PackageContextProvider>
    );

    const tocDiv = await findToc();
    const { basemapSwitcher } = await waitForBasemapSwitcher(tocDiv!);
    expect(basemapSwitcher.tagName).toBe("DIV");
});

it("should not show the basemap switcher if 'showBasemapSwitcher' is set to false", async () => {
    const { mapId, registry } = await setupMap();
    await registry.expectMapModel(mapId);
    const injectedServices = createServiceOptions({ registry });

    render(
        <PackageContextProvider services={injectedServices}>
            <Toc mapId={mapId} showBasemapSwitcher={false} data-testid="toc"></Toc>
        </PackageContextProvider>
    );

    const tocDiv = await findToc();
    const basemapSwitcher = tocDiv.querySelector(BASEMAP_SWITCHER_CLASS);
    expect(basemapSwitcher).toBeNull();
});

it("should support overriding basemap-switcher properties", async () => {
    const { mapId, registry } = await setupMap({
        layers: [
            {
                title: "OSM",
                olLayer: new TileLayer({}),
                isBaseLayer: true
            }
        ]
    });
    await registry.expectMapModel(mapId);

    const injectedServices = createServiceOptions({ registry });
    render(
        <PackageContextProvider services={injectedServices}>
            <Toc
                mapId={mapId}
                basemapSwitcherProps={{
                    allowSelectingEmptyBasemap: true,
                    className: "test-class"
                }}
                data-testid="toc"
            />
        </PackageContextProvider>
    );

    const tocDiv = await findToc();
    const { basemapSwitcher, basemapSelect } = await waitForBasemapSwitcher(tocDiv!);
    expect(basemapSwitcher?.classList.contains("test-class")).toBe(true);

    await waitFor(() => {
        const options = basemapSelect.querySelectorAll("option");
        const optionLabels = Array.from(options).map((opt) => opt.textContent);
        expect(optionLabels, "basemap options are not equal to their expected values")
            .toMatchInlineSnapshot(`
        [
          "OSM",
          "emptyBasemapLabel",
        ]
      `);
    });
});

async function waitForBasemapSwitcher(tocDiv: HTMLElement) {
    return await waitFor(() => {
        const basemapSwitcher = tocDiv.querySelector(BASEMAP_SWITCHER_CLASS);
        if (!basemapSwitcher) {
            throw new Error("basemap switcher not mounted");
        }

        const basemapSelect = basemapSwitcher?.querySelector("select");
        if (!basemapSelect) {
            throw new Error("failed to find select element in basemap switcher");
        }
        return { basemapSwitcher, basemapSelect };
    });
}

async function findToc() {
    const tocDiv = await screen.findByTestId("toc");
    return tocDiv;
}
