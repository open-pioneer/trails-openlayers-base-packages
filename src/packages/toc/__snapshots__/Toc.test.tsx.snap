// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should successfully create a toc component 1`] = `
<div
  class="toc css-2xph3x"
  data-testid="toc"
  data-theme="light"
>
  <div
    class="toc-basemap-switcher css-0"
    data-theme="light"
  >
    <h1
      class="chakra-heading css-1mnao0q"
      data-theme="light"
      id=":r0:"
    >
      basemapsLabel
    </h1>
    <div
      class="basemap-switcher css-0"
      data-theme="light"
    >
      <div
        class="basemap-switcher-select react-select--has-value css-79elbk"
        data-theme="light"
      >
        <span
          class="css-1f43avz-a11yText-A11yText"
          id="react-select-2-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
          role="log"
        />
        <div
          class="react-select__control react-select__control--menu-is-open css-156sfem"
          data-theme="light"
        >
          <div
            class="react-select__value-container react-select__value-container--has-value css-j93siq"
            data-theme="light"
          >
            <div
              class="basemap-switcher-value react-select__single-value css-1xa1gs2"
              data-theme="light"
            >
              <div
                class="css-u4p24i"
                data-theme="light"
              >
                Base layer
              </div>
            </div>
            <input
              aria-activedescendant="react-select-2-option-0"
              aria-autocomplete="list"
              aria-controls="react-select-2-listbox"
              aria-expanded="true"
              aria-haspopup="true"
              aria-labelledby=":r0:"
              aria-readonly="true"
              class="css-mohuvp-dummyInput-DummyInput"
              id="react-select-2-input"
              inputmode="none"
              role="combobox"
              tabindex="0"
              value=""
            />
          </div>
          <div
            class="react-select__indicators css-hfbj6y"
            data-theme="light"
          >
            <hr
              aria-orientation="vertical"
              class="chakra-divider react-select__indicator-separator css-1611kif"
              data-theme="light"
            />
            <div
              aria-hidden="true"
              class="react-select__indicator react-select__dropdown-indicator css-1lza9k0"
              data-theme="light"
            >
              <svg
                aria-hidden="true"
                class="chakra-icon css-onkibi"
                data-theme="light"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  fill="currentColor"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="react-select__menu css-ky7590"
          data-theme="light"
        >
          <div
            aria-multiselectable="false"
            class="react-select__menu-list css-1h5avke"
            data-theme="light"
            id="react-select-2-listbox"
            role="listbox"
          >
            <div
              aria-selected="true"
              class="basemap-switcher-option react-select__option react-select__option--is-focused react-select__option--is-selected css-e8c6zu"
              data-focus="true"
              data-theme="light"
              id="react-select-2-option-0"
              role="option"
              tabindex="-1"
            >
              <div
                class="css-u4p24i"
                data-theme="light"
              >
                Base layer
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="toc-operational-layers css-0"
    data-theme="light"
  >
    <h1
      class="chakra-heading css-1mnao0q"
      data-theme="light"
      id=":r1:"
    >
      <div
        class="css-k008qs"
        data-theme="light"
      >
        <p
          class="chakra-text css-9515vv"
          data-theme="light"
        >
          operationalLayerLabel
        </p>
        <div
          class="css-17xejub"
          data-theme="light"
        />
      </div>
    </h1>
    <ul
      aria-labelledby=":r1:"
      class="toc-layer-list css-155za0w"
      data-theme="light"
      role="list"
    >
      <li
        class="toc-layer-item layer-layer-2 css-0"
        data-theme="light"
      >
        <div
          class="toc-layer-item-content css-bk4v0k"
          data-theme="light"
        >
          <label
            class="chakra-checkbox css-1577qb8"
            data-checked=""
            data-theme="light"
          >
            <input
              aria-disabled="false"
              aria-label="Layer 2"
              class="chakra-checkbox__input"
              style="border: 0px; clip: rect(0, 0, 0, 0); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
              type="checkbox"
              value=""
            />
            <span
              aria-hidden="true"
              class="chakra-checkbox__control css-1ydjfm6"
              data-checked=""
              data-theme="light"
            >
              <div
                class="css-0"
                data-theme="light"
                style="display: flex; align-items: center; justify-content: center; height: 100%;"
              >
                <svg
                  class="css-1ckqq05"
                  data-theme="light"
                  style="fill: none; stroke-width: 2; stroke: currentColor; stroke-dasharray: 16;"
                  viewBox="0 0 12 10"
                >
                  <polyline
                    points="1.5 6 4.5 9 10.5 1"
                  />
                </svg>
              </div>
            </span>
            <span
              class="chakra-checkbox__label css-6x44c9"
              data-checked=""
              data-theme="light"
            >
              Layer 2
            </span>
          </label>
          <div
            class="css-17xejub"
            data-theme="light"
          />
        </div>
      </li>
      <li
        class="toc-layer-item layer-layer-1 css-0"
        data-theme="light"
      >
        <div
          class="toc-layer-item-content css-bk4v0k"
          data-theme="light"
        >
          <label
            class="chakra-checkbox css-1577qb8"
            data-checked=""
            data-theme="light"
          >
            <input
              aria-disabled="false"
              aria-label="Layer 1"
              class="chakra-checkbox__input"
              style="border: 0px; clip: rect(0, 0, 0, 0); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
              type="checkbox"
              value=""
            />
            <span
              aria-hidden="true"
              class="chakra-checkbox__control css-1ydjfm6"
              data-checked=""
              data-theme="light"
            >
              <div
                class="css-0"
                data-theme="light"
                style="display: flex; align-items: center; justify-content: center; height: 100%;"
              >
                <svg
                  class="css-1ckqq05"
                  data-theme="light"
                  style="fill: none; stroke-width: 2; stroke: currentColor; stroke-dasharray: 16;"
                  viewBox="0 0 12 10"
                >
                  <polyline
                    points="1.5 6 4.5 9 10.5 1"
                  />
                </svg>
              </div>
            </span>
            <span
              class="chakra-checkbox__label css-6x44c9"
              data-checked=""
              data-theme="light"
            >
              Layer 1
            </span>
          </label>
          <div
            class="css-17xejub"
            data-theme="light"
          />
        </div>
      </li>
    </ul>
  </div>
</div>
`;
